### PR TITLE
Bug fixes from incremental depmod PR

### PR DESCRIPTION
**Pull request by mwilck**
This PR splits off the bug fixes from #253, as requested by @evelikov.

I found these issues while making the kmod CI tests work with the "incremental depmod" patch set.

### CODE REVIEW

File: shared/hash.c
Comment (evelikov) [inline shared/hash.c:263]: Can we trigger this with the existing tests, alternatively let's add a test in our testsuite

File: shared/hash.c
Comment (evelikov) [inline shared/hash.c:253]: Nit: please add a blank line.

Similar to the previous commit, can we get a test. Massive thanks in advance.

File: shared/strbuf.c
Comment (evelikov) [inline shared/strbuf.c:106]: We can have this check before the reserve call.

**Review by evelikov — APPROVED** [review]
The changes look great IMHO. Would be amazing to have a test or two but it's a non-blocking request.

**Comment by mwilck** [comment]
I've added tests now that trigger the failures that I saw on #253.

I'm not sure how you deal with failures like this. I chose to add the test with `.expected_fail=true` first, and put the fix on top, reverting the `expected_fail` property.

I also added one more patch ("always use a step size of 4"), as the logic for determining the step size seems strange (I wouldn't have encountered the out-of-bounds memory access problem if the step size wasn't 1 with `n_buckets == 32`. Please have a look.

**Review by evelikov — APPROVED** [review]
You are a star, thank you for the tests - loving the `expected_fail` approach. The extra patch also looks good.

Seems like I cannot approve for CI, so it'll be up-to @lucasdemarchi

**Comment by mwilck** [comment]
rebased onto current master

**Comment by lucasdemarchi** [comment]
@evelikov  no idea why CI is not running on this PR.

**Comment by lucasdemarchi** [comment]
Thanks a lot for the fixes and additional tests. I was going to swap the order of commits and avoid using the xfail... but changed my mind - it looks good in this order. I fixed the coding style issue and applied.

**Comment by evelikov** [comment]
@lucasdemarchi some point last year, GitHub changed so that first PR from new people need to be approved by maintainers - [their blog post has the details.](https://github.blog/changelog/2023-06-06-security-enhancements-to-required-approvals-on-pull-requests/)

I suspect that adding me [as reviewer](https://github.com/orgs/community/discussions/23765#discussioncomment-8067102) might also do the trick.

**Comment by lucasdemarchi** [comment]
@evelikov the problem is that the PR itself was not showing it the approval as needed. I had to go to each individual worfklow and approve them individually.  I saw the request when the PR was opened, but it wasn't available anymore after it had been reviewed.

**Comment by lucasdemarchi** [comment]
> I suspect that adding me [as reviewer](https://github.com/orgs/community/discussions/23765#discussioncomment-8067102) might also do the trick.

it doesn't seem that there is such a role in Github. There is "triage" and "write". Anyway I created a "Reviewer" team with write access and invited both you and @stoeckmann. Note that PRs shouldn't be merged though - I manually apply the PRs pushing to both kernel.org and github.

**Pull request closed without merging**
